### PR TITLE
Move to DS 6.3

### PR DESCRIPTION
**Pull request by denisvmedyantsev**
#347

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -5,80 +5,87 @@ SHELL := /bin/bash
...
+PROJECT_PATH := /opt/savant
+
 build:
-	DOCKER_BUILDKIT=1 docker build \
-	--target base \
-	--build-arg SAVANT_VERSION=$(SAVANT_VERSION) \
-	--build-arg DEEPSTREAM_VERSION=$(DEEPSTREAM_VERSION) \
-	-f docker/$(DOCKER_FILE) \
-	-t savant-deepstream$(PLATFORM_SUFFIX) .
+	DOCKER_BUILDKIT=1 docker buildx build \
```
Comment (tomskikh) [inline Makefile]: Do we need `DOCKER_BUILDKIT=1` here? AFAIK `buildx` already uses buildkit.

File: docker/Dockerfile.deepstream [hunk 2]
```diff
@@ -58,8 +39,30 @@ COPY libs/gstsavantframemeta .
 RUN python setup.py bdist_wheel && rm -rf _skbuild
 
 
+# Savant boost(cuda) library builder
+# TODO: Update library code, CMakeLists (see build process warnings)
+# TODO: Check Jetson requirements (Xavier t194, Orin t234)
 FROM base_builder AS savantboost_build
 
+# newer cmake ver. needed for "FindCUDAToolkit"
+ARG CMAKE_VER=3.24.2
```
Comment (tomskikh) [inline docker/Dockerfile.deepstream:48]: Why cmake here was downgraded from `3.26.3` to `3.24.2`?

File: docker/Dockerfile.deepstream [hunk 3]
```diff
@@ -68,19 +71,26 @@ COPY libs/savantboost .
...
-RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 10
+
+# Savant arm64 base container
+FROM nvcr.io/nvidia/deepstream-l4t:$DEEPSTREAM_BASE_IMAGE AS base_arm64
+
+
+# Savant base container
+ARG TARGETARCH
+FROM base_${TARGETARCH} AS base
+ARG TARGETARCH
```
Comment (tomskikh) [inline docker/Dockerfile.deepstream:84]: The line is duplicated

File: docker/Dockerfile.deepstream [hunk 4]
```diff
@@ -103,36 +121,48 @@ WORKDIR $PROJECT_PATH
...
+RUN wget -qO- \
+    "https://eu-central-1.linodeobjects.com/savant-data/opencv-packages/opencv-${OPENCV_VERSION}-${TARGETARCH}-ds${DEEPSTREAM_VERSION}.tar.gz" \
+    | tar xz \
     && dpkg -i OpenCV* \
     && rm OpenCV*
 
-COPY requirements/base.txt requirements-base.txt
-RUN python -m pip install --no-cache-dir -r requirements-base.txt \
-    && rm requirements-base.txt
+COPY requirements requirements
```
Comment (tomskikh) [inline docker/Dockerfile.deepstream]: Each file from requirements was copied with a separate instruction for faster builds. E.g. when savant-rs is updated `docker build` won't install dependencies from other files again.

File: docker/Dockerfile.deepstream [hunk 5]
```diff
@@ -103,36 +121,48 @@ WORKDIR $PROJECT_PATH
...
 COPY --from=pygstsavantframemeta_build /libs/gstsavantframemeta/dist /libs/gstsavantframemeta/dist
 RUN python -m pip install --no-cache-dir /libs/gstsavantframemeta/dist/*.whl
 
-COPY --from=savant_build /tmp/build/dist /libs/savant/dist/
+COPY --from=savant_build /tmp/build/dist /libs/savant/dist
 RUN python -m pip install --no-cache-dir /libs/savant/dist/*.whl
 
+RUN rm -r /libs
+
+# remove samples to reduce image size (~1Gb)
```
Comment (tomskikh) [inline docker/Dockerfile.deepstream]: Does it affect the image size? Samples are still present in some layer anyway.

File: savant/deepstream/pipeline.py [hunk 6]
```diff
@@ -147,13 +147,16 @@ def __init__(
...
         # nvjpegdec decoder is selected in decodebin according to the rank, but
         # there are problems with the plugin:
-        # 1) https://forums.developer.nvidia.com/t/nvvideoconvert-memory-compatibility-error/226138;
-        # 2) jpeg to png conversion gives incorrect alpha channel;
-        # 3) memory type mismatch, even though we use the same
-        #  nvbuf-memory-type for nvvideoconvert and nvstreammux downstream
+        # 1) nvjpegdec doesn't support some jpg
+        #  https://forums.developer.nvidia.com/t/nvvideoconvert-memory-compatibility-error/226138;
+        # 2) nvjpegdec produces incorrect alpha channel
+        #  https://forums.developer.nvidia.com/t/nvjpegdec-produces-transparent-frames/223005
```
Comment (tomskikh) [inline savant/deepstream/pipeline.py]: Is this issue still present on DS 6.3? They said it was fixed https://forums.developer.nvidia.com/t/nvjpegdec-produces-transparent-frames/223005/16

**Review by tomskikh — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (denisvmedyantsev) [reply Makefile]: You are right. Removed.

File: docker/Dockerfile.deepstream (hunk 2 above)
Comment (denisvmedyantsev) [reply docker/Dockerfile.deepstream:48]: It doesn't matter here. I made this dockerfile from two, x86 and l4t, and used the cmake version from l4t.

File: docker/Dockerfile.deepstream (hunk 3 above)
Comment (denisvmedyantsev) [reply docker/Dockerfile.deepstream:84]: Removed

File: docker/Dockerfile.deepstream (hunk 5 above)
Comment (denisvmedyantsev) [reply docker/Dockerfile.deepstream]: No, it doesn't. Removed lines.

File: savant/deepstream/pipeline.py (hunk 6 above)
Comment (denisvmedyantsev) [reply savant/deepstream/pipeline.py]: There is only one problem out of 3 still relevant for 6.3. Fixed comment.

File: docker/Dockerfile.deepstream (hunk 4 above)
Comment (denisvmedyantsev) [reply docker/Dockerfile.deepstream]: Ok. Split it up.

**Review by tomskikh — APPROVED** [review]

**Review by abramov-oleg — APPROVED** [review]

**Pull request merged**